### DISCUSSION TITLE
Are you planning to publish Scala CLI on winget?

### DISCUSSION THREAD

**Discussion by mimoguz** [VirtusLab/scala-cli#2215] [General]
If not, do you mind if I give it a try? 

I’m transitioning from Scoop to winget for no other reason than it comes with Windows. There are some missing packages though, including Scala CLI. So I’m thinking of adding some of them myself, starting with this one.

Thanks.

**Comment by lwronski** [comment]
Hi @mimoguz, thank you for opening this discussion. It would be great to have Scala CLI on the widget. We have already opened an issue for this task at https://github.com/VirtusLab/scala-cli/issues/1181. Please let me know if I can assign you to this issue.

If you need any assistance with obtaining some an API access key or accessing any account on winget or something similar, please let us know.

**Reply by mimoguz** [reply]
> Please let me know if I can assign you to this issue.

Sure.

> If you need any assistance with obtaining some an API access key or accessing any account on winget or something similar, please let us know.

I don't think so. Adding a package to the community repository seems straitforward:

> To submit a package to this repository, you should follow these steps:
> 
> 1. Follow the [Contributing](#Contributing) guidelines below.
> 2. [Author](AUTHORING_MANIFESTS.md) a Manifest.
> 3. [Test](#test-your-manifest) your manifest.
> 4. [Submit](#submit-your-pr) your pull request (PR).
> 5. Respond to any feedback in your PR.

This is my fork: [Github link](https://github.com/mimoguz/winget-pkgs/tree/de6651507d9efa2344817e93b43fac4583807974/manifests/v/VirtusLab/ScalaCLI/1.0.1). Please let me know if it's O.K.

**Reply by lwronski** [reply]
It's fine for me. I have a question about updating. When Scala CLI is released in `1.0.2`, what should be done? Do we need to open a PR to winget with a new Manifest, or does winget have a mechanism to detect automatically when Scala CLI is released in GitHub?

**Reply by mimoguz** [reply]
I think a new PR is required, since most of the PRs are for updates. But maybe you can integrate winget-create to your build pipeline:
[https://github.com/microsoft/winget-create/blob/main/doc/update.md](https://github.com/microsoft/winget-create/blob/main/doc/update.md).

**Reply by lwronski** [reply]
One drawback of this solution is that we need to generate a personal access token from GitHub and keep it in the scala-cli repository. But despite this, seems to me that you can open a PR to the main Windows Package Manager repo with the new Scala CLI manifest.

Let me know when it's merged, then we need to update the documentation.

**Reply by mimoguz** [reply]
> Let me know when it's merged, then we need to update the documentation.

Sure. 

[The PR](https://github.com/microsoft/winget-pkgs/pull/110242)

**Reply by mimoguz** [reply]
> One drawback of this solution is that we need to generate a personal access token from GitHub and keep it in the scala-cli repository. But despite this, seems to me that you can open a PR to the main Windows Package Manager repo with the new Scala CLI manifest.

Quoting [https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens):

> When using a personal access token in a GitHub Actions workflow, consider whether you can use the built-in GITHUB_TOKEN instead. For more information, see "[Automatic token authentication](https://docs.github.com/en/actions/security-guides/automatic-token-authentication)." 

Maybe this can help if you decide to automate.

**Reply by lwronski** [reply]
I'm not sure if the Automatic token `GITHUB_TOKEN` grants permission to open a new PR in another repository.

**Reply by mimoguz** [reply]
> I'm not sure if the Automatic token GITHUB_TOKEN grants permission to open a new PR in another repository.

:( 

This is sbt's winget workflow btw: [https://github.com/sbt/sbt/blob/1.9.x/.github/workflows/winget.yml](https://github.com/sbt/sbt/blob/1.9.x/.github/workflows/winget.yml)

Anyway I can submit update PRs if short delays won't be a problem.

**Reply by lwronski** [reply]
> Anyway I can submit update PRs if short delays won't be a problem.

In general, I think it would be better to automate the process within the release CI. From VirtusLab, we can generate a token that can be used when running `winget upgrade`. I suggest adding new steps in the release CI of scala-cli, after the [last step](https://github.com/VirtusLab/scala-cli/blob/c8132b58c9469ed02a7470a796a808c8029d9feb/.github/workflows/ci.yml#L1061) of publishing the Chocolatey package which this step publish Scala CLI on winget.

**Reply by mimoguz** [reply]
Then using [WinGet Releaser](https://github.com/vedantmgoyal2009/winget-releaser), I think adding this step would be enough:

```yaml
- uses: vedantmgoyal2009/winget-releaser@v2
        with:
          identifier: VirtusLab.ScalaCLI
          installers-regex: '\.msi$'
          token: ${{ secrets.WINGET_TOKEN }}
```

**Reply by lwronski** [reply]
@mimoguz, Thank you for preparing the steps. You are welcome to open a pull request with the changes.

**Reply by mimoguz** [reply]
> You are welcome to open a pull request with the changes.

Are you asking me to break your CI? 😯😀 Sure, I can do that. I don't have the faintest idea how to test that though.

**Reply by lwronski** [reply]
I don't have any ideas on how to test it either. I think the only way to verify it is by testing it in the next release of scala-cli. If something doesn't work as expected, we can still fix it manually. Additionally, this step would be the last one in the CI release, ensuring that it doesn't break any other steps.

**Reply by mimoguz** [reply]
So, I guess it didn't work?

**Reply by lwronski** [reply]
I'm not sure  about the root cause of error in https://github.com/VirtusLab/scala-cli/actions/runs/5520594036/jobs/10072534547. The action seems unable to fork from `scala-steward/winget-pkgs`, even though the repository exists at https://github.com/scala-steward/winget-pkgs.

**Reply by lwronski** [reply]
We fixed it and not it works - https://github.com/microsoft/winget-pkgs/pull/111913.

**Reply by mimoguz** [reply]
👍🏽

May I ask what the issue was?

**Reply by lwronski** [reply]
I need to add option `fork-user:` to  `vedantmgoyal2009/winget-releaser@v2`, because fork  exists in `scala-steward` instead of `virtuslab` user.

**Reply by lwronski** [reply]
Thank you again for your help, @mimoguz. From the next release, this should work automatically.

**Comment by mimoguz** [comment]
PR was merged 🎉

![scalacli](https://github.com/VirtusLab/scala-cli/assets/7283047/42d7d05f-ad4e-4f24-984f-c982dfaca1e8)

**Reply by lwronski** [reply]
@mimoguz Thank you!
